File: interfaceTest/ssros_ams/src/set_fan_pwm_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "test_interfaces/srv/set_fan_pwm.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  if (argc != 2) {
      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "usage: set_fan_pwm_client PWM");
      return 1;
  }

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("set_fan_pwm_client");
  rclcpp::Client<test_interfaces::srv::SetFanPWM>::SharedPtr client =
    node->create_client<test_interfaces::srv::SetFanPWM>("set_fan_pwm");

  auto request = std::make_shared<test_interfaces::srv::SetFanPWM::Request>();
  request->pwm = atoll(argv[1]);

  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "service not available, waiting again...");
  }

  auto result = client->async_send_request(request);
  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Success: %d", result.get()->success);
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Failed to call service set_fan_pwm");
  }

  rclcpp::shutdown();
  return 0;
}

File: interfaceTest/ams/ssros_ams_py/ssros_ams_py/set_fan_pwm_client.py
import sys
from test_interfaces.srv import SetFanPWM
import rclpy
from rclpy.node import Node


class SetFanPWMClient(Node):
    def __init__(self):
        super().__init__("set_fan_pwm_client_py")
        self.cli = self.create_client(SetFanPWM, "set_fan_pwm")
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("service not available, waiting again...")
        self.req = SetFanPWM.Request()

    def send_request(self):
        self.req.pwm = int(sys.argv[1])
        self.future = self.cli.call_async(self.req)


def main(args=None):
    rclpy.init(args=args)

    if len(sys.argv) != 2:
        print("usage: set_fan_pwm_client PWM")
        sys.exit(1)

    set_fan_pwm_client = SetFanPWMClient()
    set_fan_pwm_client.send_request()

    while rclpy.ok():
        rclpy.spin_once(set_fan_pwm_client)
        if set_fan_pwm_client.future.done():
            try:
                response = set_fan_pwm_client.future.result()
            except Exception as e:
                set_fan_pwm_client.get_logger().info("Service call failed %r" % (e,))
            else:
                set_fan_pwm_client.get_logger().info(
                    "Result of set_fan_pwm: for %d, success: %d"
                    % (set_fan_pwm_client.req.pwm, response.success)
                )
            break

    set_fan_pwm_client.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
